### ISSUE TITLE
[RFC] What is the integer overflow story in Julia?

### ISSUE THREAD

**Issue by LilithHafner** [JuliaLang/julia#50486]
For simple operations (e.g. `+`, `*`, `-`) we do modular arithmetic and do not warn on overflow.
For complex operations, we sometimes check for intermediate overflow ([`lcm`](33781), [`getindex`](50118), `binomial`) and sometimes do not, ([`isapprox`](50380), [`lcm`](33926))

I'd like a clear, documented, policy of when we let overflow happen, when we throw, and when we make sure to avoid it with widen or similar.

The first step is to come up with the policy. Ideas?

**Comment by stevengj** [comment]
See also #21600.

**Comment by JeffreySarnoff** [comment]
Regarding a policy for the treatment of integer overflow.

It is easier to make the case for treating unsigned bitstypes as modular values (permitting operations to wraparound either way) than to make a compelling case for treating signed bitstypes as modular values. Many uses of unsigned bitstypes are designed with a reliance on the presence of unchecked wraparound. 

It is rare to design an algorithm that requires unchecked wraparound for signed bitstypes.  The only common rationale for allowing signed bitstypes to wrap (over or under) is that provides accelerated throughput and the associated simplification of instruction flow is easier to optimize.

[Integer Wraparound](https://cwe.mitre.org/data/definitions/190.html) is #14 on the 2023 CWE list of Top 25 Most Dangerous Software Weaknesses

> CWE™ is a community-developed list of software and hardware weakness types. It serves as a common language, a measuring stick for security tools, and as a baseline for weakness identification, mitigation, and prevention efforts.
(The Mitre Corporation underwrites CWE™)

**Comment by nhz2** [comment]
I think of `Int64` as the 64 least significant bits of a hypothetical `BigInt` sign extended.

Therefore, I believe the following equality should hold for all basic integer operations, or they should
throw an error or have some warning in their docstring.
```julia
op(x::Int64, y::Int64) == op(big(x), big(y)) % Int64
```

For example, one case where the above equality doesn't work is with `^`:

`1 ^ -2 == 1.0`
`Int64(1) ^ Int64(-2) == 1`
 `big(1) ^ big(-2)` throws a DomainError.

But I guess this is okay because the docstring for `^` says it has special behavior when using `Int` literals.

Though I'm not sure why  `big(1) ^ big(-2)` errors here.

**Comment by PallHaraldsson** [comment]
>one case where the above equality doesn't work is with ^:
>
> 1 ^ -2 == 1.0
> Int64(1) ^ Int64(-2) == 1

That's because arguably Integer ^ Integer should in general be Float64 (if not Decimal), one of few flaws in Julia (only exception when b of a ^ b is Unsigned, then I can argue for status quo), for the same reason Integer / Integer is Float64, which is a good decision. 

Such a decision would have averted overflow or costly check (and is only done at parser time for constants), and I would like to see it changed for Julia 2.0 (see other issue on it, and what can be done for 1.x, let's not discuss this here). Insisting on Int result (for Signed b), is arguably faulty math, and I don't think many would rely on that very much. This is one reason I want 2.0, to break this (but in a good way), most would be better off, I don't think ^ is very much used anyway for integers except for Unsigned in number theory and wouldn't be affected.

>big(1) ^ big(-2) throws a DomainError.

Since negative power gives a fraction, in general not an integer, why you should do (and only the above for positive power):
>julia> BigFloat(1) ^ BigFloat(-2)
>1.0

**Comment by adienes** [comment]
I would say basically only arithmetic or bit-twiddling operators should allow unchecked overflow; all other functions should check internally when there is danger

**Comment by adienes** [comment]
probably to the annoyance of viewers of the issue tracker, I've been finding & attempting to fix several bugs related to internal overflow lately, and here's my latest idea for an extremely minimal documented policy:

suppose there is some function `foo: T -> T` so for simplicity just say `T` is some integral domain and `foo` is type stable. and say it has some potential for overflow

```
function foo(input)
    intermediate = bar(input)
    return final(intermediate)
end
```

and the "correct" behavior is described in the documentation of `foo` by some statement `F`. there may be some inputs `x` where due to internal overflow, it is difficult or cumbersome or simply not possible to make `foo(x) == F(x)`. I propose:

* documentation is _mandatory_ if `foo(x) != F(x)` without error.
* throwing an `OverflowError` or falling back to a slower implementation is _always_ allowed
* if `foo(x) != F(x)` without a specific note in the docs for `foo`, that is considered a bug
* type stability should generally be assumed, so if `foo` chooses to widen as its "slow fallback" this should be documented

I think this is actually pretty close to empirical status quo, since it allows for many judgement calls about when to error vs be slow vs just return the wrong thing. but what it changes is when the wrong thing can be returned this must be documented. All arithmetic operations already have such a warning in the documentation, and I don't think there are actually all that many more functions that intentionally behave this way (although there are some)

applying the example to some open issues / PRs:

* https://github.com/JuliaLang/julia/pull/59678 `gcd` family chooses `Error` (not always `OverflowError`, but close enough) so no documentation needed
* https://github.com/JuliaLang/julia/issues/45880#issuecomment-1171647181 `nextpow2` needs to be fast, so it wants license to return the wrong thing. this is fine but must be documented
* https://github.com/JuliaLang/julia/issues/50930 `DateTime` should choose one of slow fallback or error, as it's rarely performance critical


if there is support for this policy I'm happy to make a PR to the docs. obviously it doesn't even address when to widen vs error vs documentation hazard sign, and I don't know if there really can be a unified / consistent / generic policy for that since it's so context-dependent, but I think this is at least better than nothing.

**Comment by LilithHafner** [comment]
In the case where `foo(x) != F(x)`, I see a distinction based on weather `F(x)` is representable as an object of type `T`. If it is, then we should strongly prefer to return `F(x)` and anything else is a bug (or should be documented, therefore redefining `F`). In the case where `F(x)` is not representable (e.g.` *(::Int, ::Int)`), returning `F(x)%T` for bit integer types is very standard and may warrant a mention but not a warning in each docstring.

TL;DR Overflow in internals is distinctly worse than rounding/truncating/coercing the final result to fit into a type stable return type.
